schema/schema.go

```diff
@@ -0,0 +1,215 @@
+	schema "github.com/xeipuuv/gojsonschema"
+)
+
+const (
+	// BuiltinSchemaName references the builtin schema for Load()/Set().
+	BuiltinSchemaName = "builtin"
+	// DefaultSchemaName is the builtin schema.
+	DefaultSchemaName = BuiltinSchemaName
+	// NoneSchemaName references a disabled/NOP schema for Load()/Set().
+	NoneSchemaName = "none"
+
+	// builtinSchemaFile is the builtin schema URI in our embedded FS.
+	builtinSchemaFile = "file:///schema.json"
+)
+
+// Schema is a JSON validation schema.
+type Schema struct {
+	schema *schema.Schema
+}
+
+// Error wraps a JSON validation result.
+type Error struct {
+	Result *schema.Result
+}
+
+// Set sets the default validating JSON schema.
+func Set(s *Schema) error {
+	current = s
+	return nil
+}
+
+// Get returns the active validating JSON schema.
+func Get() *Schema {
+	return current
+}
+
+// BuiltinSchema returns the builtin validating JSON Schema.
+func BuiltinSchema() *Schema {
+	return builtin
+}
+
+// NopSchema returns an validating JSON Schema that does no real validation.
+func NopSchema() *Schema {
+	return &Schema{}
+}
+
+// ValidateData validates the given JSON document against the active schema.
+func ValidateData(data []byte) (error, error) {
+	return current.ValidateData(data)
+}
+
+// ValidateFile validates the given JSON file against the active schema.
+func ValidateFile(path string) (error, error) {
+	return current.ValidateFile(path)
+}
+
+// Load the given JSON Schema.
+func Load(source string) (*Schema, error) {
+	var (
+		loader schema.JSONLoader
+		err    error
+		s      *schema.Schema
+	)
+
+	source = strings.TrimSpace(source)
+
+	switch {
+	case source == BuiltinSchemaName:
+		return BuiltinSchema(), nil
+	case source == NoneSchemaName, source == "":
+		return NopSchema(), nil
+	case strings.HasPrefix(source, "file://"):
+	case strings.HasPrefix(source, "http://"):
+	case strings.HasPrefix(source, "https://"):
+	default:
+		if strings.Index(source, "://") < 0 {
+			source, err = filepath.Abs(source)
+			if err != nil {
+				return nil, err
+			}
+			source = "file://" + source
+		}
+	}
+
+	loader = schema.NewReferenceLoader(source)
+
+	s, err = schema.NewSchema(loader)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to load JSON validation schema")
+	}
+
+	return &Schema{schema: s}, nil
+}
+
+// ValidateData validates the given JSON data against the schema.
+func (s *Schema) ValidateData(data []byte) (error, error) {
+	var (
+		any interface{}
+		err error
+	)
+
+	if !strings.HasPrefix(strings.TrimSpace(string(data)), "{") {
+		err = yaml.Unmarshal(data, &any)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed unmarshal as YAML")
+		}
+		data, err = json.Marshal(any)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed remarshal as JSON")
+		}
+	}
+
+	return s.validate(schema.NewStringLoader(string(data)))
+}
+
+// ValidateFile validates the given JSON file against the schema.
+func (s *Schema) ValidateFile(path string) (error, error) {
```

Comment:
It might be useful to add a `Validate` method that accepts an `io.Reader` which is simpler to create from both paths and byte slices.

Reply:
Good point. I try to roll such an interface.

Reply:
I added such an interface.